src/gluonnlp/data/data_preprocessing_transform.py

```diff
@@ -0,0 +1,515 @@
+from .qa_preprocessing_utils import truncate_seqs_equal, improve_answer_span
+
+
+class TruncateTransform:
+    """
+    Truncate a sequence(list) to max length.
+
+    Parameters
+    ----------
+    max_len : int
+    truncate_fn : callable
+        A function determines how to truncate the sequence (list).
+        The function should implement argument max_length
+
+    Returns
+    -------
+    list : list of sequences or a single sequence
+    """
+    def __init__(self, max_len, truncate_fn=truncate_seqs_equal):
+        self._max_len = max_len
+        self.fn = partial(truncate_fn, max_length=max_len)
+
+    def __call__(self, seqs):
+        assert isinstance(seqs, collections.abc.Iterable)
+        if len(seqs) == 0:
+            return seqs
+        if isinstance(seqs[0], collections.abc.Iterable) and not isinstance(
+                seqs[0], str):
+            #if it contains a list of seqs
+            seqs = self.fn(seqs)
+            return seqs
+        else:
+            #if it is a single sequence
+            return seqs[:self._max_len]
+
+
+class ConcatSeqTransform:
+    """Insert special tokens for sequence list or a single sequence.
+    For sequence pairs, the input is a list of 2 strings:
+    text_a, text_b.
+
+    Inputs:
+        text_a: 'is this jacksonville ?'
+        text_b: 'no it is not'
+        start_token: [CLS]
+        token_after_seg: [[SEP], [SEP]]
+        end_token: None
+
+    Processed:
+        tokens: '[CLS] is this jacksonville ? [SEP] no it is not . [SEP]'
+        segment_ids: 0 0  0    0            0  0    1  1  1  1   1 1
+        p_mask:      0 0  0    0            0  1    0  0  0  0   0 1
+        valid_length: 12
+
+    Parameters
+    ----------
+    vocab : Vocab
+        If vocab is not None. The tokens will be converted to ids before return
+
+    token_after_seg : list
+        The special tokens to be appended to each sequence. For example:
+        Given:
+            seqs: [[1, 2], [3, 4], [5, 6]]
+            token_after_seg: [None, 7]
+        it will be:
+            [1, 2, 3, 4, 7, 5, 6]
+
+    start_token : string
+        The special token to be added to the start
+
+    end_token : string
+        The special token to be added to the end
+
+    seqs : list of sequences or a single sequence
+
+    Returns
+    -------
+    np.array: input token ids in 'int32', shape (batch_size, seq_length)
+    np.array: segment ids in 'int32', shape (batch_size, seq_length)
+    np.array: valid length in 'int32', shape (batch_size,)
+    np.array: mask for special tokens
+    """
+    def __init__(self,
+                 vocab=None,
+                 token_after_seg=None,
+                 start_token=None,
+                 end_token=None):
+        self._vocab = vocab
+        self._start_token = start_token
+        self._end_token = end_token
+        self._token_after_seg = token_after_seg if token_after_seg else []
+
+    def __call__(self, seqs):
```

Comment:
The implementation looks quite complex to me. I think you can simplify it by seeing this operator as a kind of `zip` operator, which is zipping the sequences and the seperators together.
Consider you expect the following input format:

``` python
seperators = [[], [7], [7]]
sequences = [[1, 2], [3, 4], [5, 6]]
```

Then you can calculate the 3 quantities with relative ease based on `sum` and `itertools.zip_longest`. 

``` python
concat = sum((sep + seq for sep, seq in itertools.zip_longest(seperators, sequences, fillvalue=[])), [])
segment_ids = sum(([i] * (len(sep) + len(seq)) for i, (sep, seq) in enumerate(itertools.zip_longest(seperators, sequences, fillvalue=[]))), [])
p_mask = sum(([1] * len(sep) + [0] * len(seq) for i, (sep, seq) in enumerate(itertools.zip_longest(seperators, sequences, fillvalue=[]))), [])
```

In the example above, the output is

``` python
concat = [1, 2, 7, 3, 4, 7, 5, 6]
segment_ids = [0, 0, 1, 1, 1, 2, 2, 2]
p_mask = [0, 0, 1, 0, 0, 1, 0, 0]
```

But users can also provide more complex inputs, such as

``` python
seperators = [[0, 0], [7], [7], [1, 1]]
sequences = [[1, 2], [3, 4], [5, 6]]
```

resulting in 

``` python
concat = [0, 0, 1, 2, 7, 3, 4, 7, 5, 6, 1, 1]
segment_ids = [0, 0, 0, 0, 1, 1, 1, 2, 2, 2, 3, 3]
p_mask = [1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1]
```